x-pack/apm-server/aggregation/txmetrics/aggregator.go

```diff
@@ -293,49 +332,106 @@ func (a *Aggregator) updateTransactionMetrics(key transactionAggregationKey, has
 
 	m := a.active
 	m.mu.RLock()
-	entries, ok := m.m[hash]
+	var ok bool
+	var entries []*metricsMapEntry
+	svcEntry, svcOk := m.m[key.serviceName]
+	if svcOk {
+		entries, ok = svcEntry.m[hash]
+	}
 	m.mu.RUnlock()
 	var offset int
 	if ok {
 		for offset = range entries {
 			if entries[offset].transactionAggregationKey.equal(key) {
 				entries[offset].recordDuration(duration, count)
-				return true
+				return
 			}
 		}
 		offset++ // where to start searching with the write lock below
 	}
 
+	entries = nil
+	var svcOverflow bool
 	m.mu.Lock()
-	entries, ok = m.m[hash]
-	if ok {
-		for i := range entries[offset:] {
-			if entries[offset+i].transactionAggregationKey.equal(key) {
-				m.mu.Unlock()
-				entries[offset+i].recordDuration(duration, count)
-				return true
+	svcEntry, svcOk = m.m[key.serviceName]
```

Comment:
I think we can remove all the `m.mu.Unlock()` calls below and just add a defer statement to unlock to simplify the locking in this function.
```suggestion
	defer m.mu.Unlock()
	svcEntry, svcOk = m.m[key.serviceName]
```

Reply:
Thanks for this. I deferred doing this to later since the earlier code was keeping the `recordDuration` out of the critical section so I wanted to check if there are any other edge cases but forgot to do it.